core/math/quaternion.cpp

```diff
@@ -189,16 +195,70 @@ Quaternion Quaternion::slerpni(const Quaternion &p_to, const real_t &p_weight) c
 			invFactor * from.w + newFactor * p_to.w);
 }
 
-Quaternion Quaternion::cubic_slerp(const Quaternion &p_b, const Quaternion &p_pre_a, const Quaternion &p_post_b, const real_t &p_weight) const {
-#ifdef MATH_CHECKS
-	ERR_FAIL_COND_V_MSG(!is_normalized(), Quaternion(), "The start quaternion must be normalized.");
-	ERR_FAIL_COND_V_MSG(!p_b.is_normalized(), Quaternion(), "The end quaternion must be normalized.");
+static real_t cubic_interpolate_real(const real_t p_pre_a, const real_t p_a, const real_t p_b, const real_t p_post_b, real_t p_c) {
+	// This is cloned straight from animation.cpp
+
+	real_t p0 = p_pre_a;
+	real_t p1 = p_a;
+	real_t p2 = p_b;
+	real_t p3 = p_post_b;
+
+	real_t t = p_c;
+	real_t t2 = t * t;
+	real_t t3 = t2 * t;
+
+	return 0.5f *
+			((p1 * 2.0f) +
+					(-p0 + p2) * t +
+					(2.0f * p0 - 5.0f * p1 + 4.0f * p2 - p3) * t2 +
+					(-p0 + 3.0f * p1 - 3.0f * p2 + p3) * t3);
+}
+
+Quaternion Quaternion::cubic_interpolate(const Quaternion &p_q, const Quaternion &p_prep, const Quaternion &p_postq, const real_t &p_t, const bool flip_to_shortest_path) const {
+	Quaternion q0;
+	Quaternion q1;
+	Quaternion q2;
+	Quaternion q3;
+
+	if (flip_to_shortest_path) {
+		// Flip quaternions to shortest path if necessary (this used as the reference (never flipped))
+		q0 = this->dot(p_prep) < 0.0f ? -p_prep : p_prep;
+		q1 = *this;
+		if (this->dot(p_q) < 0.0f) {
+			q2 = -p_q;
+			// When comparing with already flipped q2, we need to flip q3 also when dot product is 0
+			// This keeps flipping between subsequent quaternions consistent.
+			q3 = q2.dot(p_postq) <= 0.0f ? -p_postq : p_postq;
+		} else {
+			q2 = p_q;
+			q3 = q2.dot(p_postq) < 0.0f ? -p_postq : p_postq;
+		}
+	} else {
+		q0 = p_prep;
+		q1 = *this;
+		q2 = p_q;
+		q3 = p_postq;
+	}
+
+	// TODO: Remove test output
```

Comment:
Is this todo still valid?

Reply:
I'd like to remove the test output code completely (instead of disabling it using #if) finally, but left this for testing purposes for now. I will remove it if this method is decided to be merged.